lex/outputbis.ml

```diff
@@ -226,15 +316,49 @@ let output_lexdef ic oc tr header rh
 
   copy_chunk ic oc tr header false;
   let has_refill = output_refill_handler ic oc tr rh in
-  output_automata oc has_refill transitions;
+  let counters = Array.make (Array.length transitions) 0 in
+  let count i = counters.(i) <- counters.(i) + 1 in
+  List.iter (fun e -> count (fst e.auto_initial_state)) entry_points;
+  Array.iter
+    (function
+      | Shift (_, a) ->
+          let tbl = Hashtbl.create 8 in
+          Array.iter
+            (function
+              | (Goto i, _) when not (Hashtbl.mem tbl i) ->
+                  Hashtbl.add tbl i (); count i
+              | _ -> ()
+            )
+            a
+      | Perform _ -> ()
+    )
+    transitions;
+  let inline =
+    Array.mapi
+      (fun i -> function
+         | Perform _ -> true
+         | Shift _ -> counters.(i) = 1
+      )
+      transitions
+  in
+  let goto_state ctx n =
+    if inline.(n) then
+      output_trans_body ctx transitions.(n)
+    else
+      pr ctx "    __ocaml_lex_state%d lexbuf %s _buf _len _curr _last %s\n" n
+        (last_action ctx)
+        (if ctx.has_refill then " k" else "")
+  in
+  let ctx = {has_refill; oc; goto_state; last_action=None} in
```

Comment:
I think it would be nicer to move this big blob of code to its own auxiliary function (the current proposal is "mixing abstraction levels").

Reply:
Is the new code better?  Another approach would be to avoid the first-class function in the "context", but this would force to make many functions mutually recursive.